rust/gui-client/src-tauri/src/client/gui.rs

```diff
@@ -427,56 +429,64 @@ pub(crate) enum ControllerRequest {
     UpdateNotificationClicked(Url),
 }
 
+enum Status {
+    /// Firezone is disconnected.
+    Disconnected,
+    /// Firezone is signing in and raising the tunnel.
+    Connecting,
+    /// Firezone is ready to use.
+    TunnelReady,
+}
+
+impl Default for Status {
+    fn default() -> Self {
+        Self::Disconnected
+    }
+}
+
+impl Status {
+    /// Returns true if connlib has started, even if it's still signing in.
+    fn connlib_is_up(&self) -> bool {
+        match self {
+            Self::Disconnected => false,
+            Self::Connecting => true,
+            Self::TunnelReady => true,
+        }
+    }
+}
+
 struct Controller {
     /// Debugging-only settings like API URL, auth URL, log filter
     advanced_settings: AdvancedSettings,
     app: tauri::AppHandle,
     // Sign-in state with the portal / deep links
     auth: client::auth::Auth,
     ctlr_tx: CtlrTx,
-    /// connlib session for the currently signed-in user, if there is one
-    session: Option<Session>,
+    ipc_client: ipc::Client,
     log_filter_reloader: logging::Reloader,
     /// Tells us when to wake up and look for a new resource list. Tokio docs say that memory reads and writes are synchronized when notifying, so we don't need an extra mutex on the resources.
     notify_controller: Arc<Notify>,
-    tunnel_ready: bool,
+    resources: Arc<ArcSwap<Vec<ResourceDescription>>>,
+    status: Status,
     uptime: client::uptime::Tracker,
 }
 
-/// Everything related to a signed-in user session
-struct Session {
-    callback_handler: CallbackHandler,
-    connlib: ipc::Client,
-}
-
 impl Controller {
     /// Pre-req: the auth module must be signed in
     async fn start_session(&mut self, token: SecretString) -> Result<(), Error> {
-        if self.session.is_some() {
-            Err(anyhow!("can't start session, we're already in a session"))?;
+        if self.status.connlib_is_up() {
+            Err(anyhow::anyhow!(
+                "Can't connect to Firezone, we're already connected."
+            ))?;
```

Comment:
Would there be a benefit in making this idempotent instead? Something more like a behaviour of "set new token".

Another thought I've been mulling over is that the behaviour of reading the secret string from the keyring is something the background service could do itself. Then the GUI would only have to send a notify message that the token in the key-ring changed (after authenticating) for the service to reload it.

Reply:
Yeah this should be idempotent. I think I was imagining that if the user clicks "Sign In" when we're already signed in, it must be a logic error, but the whole auth flow happens between "Sign In" and `start_session` so that error can go somewhere else if it's still needed

> the behaviour of reading the secret string from the keyring is something the background service could do itself.

Probably not on Linux, since we rely on the desktop's keyring, and the IPC service doesn't get access to a desktop.

I think on both OSes, the user's login password is stretched into a key to encrypt credentials, which protects the token from being read off the hard drive when the computer is shut down. _Maybe_ on newer Windows it can use the TPM to provide encryption for system service accounts, but I don't know how that works, and I doubt it works on Linux.

That wouldn't be a problem if we ran the IPC service under the user's account, but then getting root privilege on Linux to control DNS would be hard.

So maybe, but I'd have to understand more about security to be sure about it.

Reply:
Actually looking at the code, I don't think there is a correct way to implement that, because `Session::disconnect` is (similar to reconnect and set_dns and others) fire-and-forget, there's no sync or async way to wait for the connlib task and worker threads to exit https://github.com/firezone/firezone/blob/663367b6055ced7432866a40a60f9525db13288b/rust/connlib/clients/shared/src/lib.rs#L98-L103